_rules/aria-required-context-role-ff89c9.md

```diff
@@ -0,0 +1,257 @@
+---
+id: ff89c9
+name: ARIA required context role
+rule_type: atomic
+description: | 
+  This rule checks that a role exists inside its required context.
+accessibility_requirements:
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects: 
+- DOM Tree
+- CSS Styling
+authors:
+- Anne Thyme Nørregaard
+---
+
+## Applicability
+
+The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has an explicit [semantic role](#semantic-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an implicit semantic role that is identical to its explicit semantic role.
+
+**Note:** An example of an element that has an implicit semantic role that is identical to its explicit semantic role is an `<li>` element that has `role="listitem"`. This rule is not applicable to elements like this.
+
+## Expectation
+
+The target element is [owned by](#owned-by) an element that has a [semantic role](#semantic-role) that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element.
```

Comment:
We should also allow roles that inherit from the required context. Romain came up with this one a few days ago. You could put this as an passed example:

```html
<div role="dictionary">
  <div role="listitem">Foo</div>
</div>
```

Reply:
@WilcoFiers, as I understand, what you are asking for is to change the applicability to something like: 

> The target element is [owned by](#owned-by) an element that has a [semantic role](#semantic-role) that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element, or that inherits from the one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element.

... where "inherits from a role" means "has the role in question listed as a [superclass role](https://www.w3.org/TR/wai-aria-1.1/#superclassrole) (and the roles that inherits can reversely been found listed as [subclass roles](https://www.w3.org/TR/wai-aria-1.1/#subclassroles) for the [superclass role](https://www.w3.org/TR/wai-aria-1.1/#superclassrole)."

I don't know if I agree with this. 
As long as we are working within the WAI-ARIA spec (and we are in this rule, according to the expectation that says "one of the WAI-ARIA required context roles), both superclass and subclass roles will be listed as possible required context roles for a role. See e.g. [`row` (role)](https://www.w3.org/TR/wai-aria-1.1/#row) that both lists [`grid` (role)](https://www.w3.org/TR/wai-aria-1.1/#grid) and [`table` (role)](https://www.w3.org/TR/wai-aria-1.1/#table), even though `grid` is a subclass of the `table` role.

I am not sure where you are getting the `role="dictionary"` from? It's not in the WAI-ARIA spec.
Maybe you are thinking about [`doc-glossary` role](https://www.w3.org/TR/dpub-aria-1.0/#doc-glossary) from DPUB-ARIA that has `term` and `definition` as required owned elements? Since that just has the superclass role of `landmark`, adding something on inheritance wouldn't help anything here. 
And then there is also the fact that this is in another standard than the one used in the expectation...

Reply:
@WilcoFiers, I have added a note on the superclass and subclass roles being listed as required context roles in WAI-ARIA. 
Is this, along with my previous comment, enough to handle your concern?

Reply:
Hey Anne. Apologies, I messed up my example. What I was thinking of is `role="directory"`, or similarly, `role="feed"`. Neither of those are listed as required context roles for `listitem`.  I don't know if that's an oversight in the ARIA spec or not. DPUB-ARIA may also be something worth considering. It does something quite unusual, like this:

```html
<section role="doc-bibliography">
   <h1>Cited Works</h1>
   <div role="list">
      <p role="doc-biblioentry" id="b8cab5dd-bc24-459c-9858-7afa9da69b64">
         John Steinbeck, The Grapes of Wrath (New York: The Viking Press, 1939)
      </p>
   </div>
   …
</section>
```

In DPUB-ARIA, `doc-biblioentry` has a required context role of `doc-bibliography`. The way we've defined this rule, this example, straight out of DPUB-ARIA, would fail. My guess is that the DPUB-ARIA folks actually took the literal "owned elements" definition from ARIA (the one we established is wrong), and went with that. I'm not too sure what to do with that. Perhaps @rdeltour has any ideas?

Reply:
It's always been my interpretation that the required context role included the stated role **and all its subclass roles**, but I'm no RDF expert. So yes, my take is that `listitem` can definitely be owned by `directory`. We should ask the ARIA folks for a clarification.

> My guess is that the DPUB-ARIA folks actually took the literal "owned elements" definition from ARIA (the one we established is wrong),

Can you please clarify (or give a pointer to) what do you mean by the "literal owned elements definition", and what was it that you established was wrong?

/cc @mattgarrish

Reply:
FYI I just filed w3c/dpub-aria#15

Reply:
Thanks for getting back, @WilcoFiers and @rdeltour .

When looking in the WAI-ARIA spec for mentions of `directory`, I think it becomes pretty clear that it is missing in the list of required context roles for `listitem`.

E.g.: 

> `listitem` (role) 
> A single item in a list or directory.

It just seems like it has slipped from the table, because it is mentioned in paragraphs in different places.

Looking further into it, `list` also has the subclass of [`feed`](https://www.w3.org/TR/wai-aria-1.1/#feed), which is...

> A scrollable list of articles where scrolling may cause articles to be added to or removed from either end of the list.

In this case, I do not think `feed` should be considered an allowed context role for `listitem`, even though it is a subclass of `list`, which to me indicates that only the subclasses that are actually listed as required context roles should be considered.

Reply:
Suggestion from call between Wilco, Romain, Kasper and Anne:

"... that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element, or a DPUB subclass of one the required context roles."

Add note about "directory" being incorrect.

Reply:
Raised an issue for the WAI-ARIA group on the `directory` (role) missing as a required context role for `listitem`: https://github.com/w3c/aria/issues/1030